### PR TITLE
Support, openapi spec added

### LINKED ISSUE #343: feat: add support of OpenAPI spec that ajv do not initially support
## 🚀 Feature Proposal

Adding the support for following types.
```
{ type: 'string', format: 'binary' }
{ type: 'string', format: 'byte' }
{ type: 'number', format: 'int32' }
{ type: 'number', format: 'int64' }
```
Example for validator can be found in here, [ajv-oai](https://github.com/oaijs/ajv-oai) or [ajv-openapi](https://github.com/touchifyapp/ajv-openapi)

## Motivation

Probably solving all the related issues #47

## Example

As `fastify` itself do not provide a way to `addFormat` in `ajv`. I mange to provide the user a custom `ajv` instance or `validatorCompiler` function which prefixed the additional fomat.

```TS
import Ajv from 'ajv'
export const ajv = new Ajv({
  removeAdditional: true, // remove additional properties
  useDefaults: true, // replace missing properties and items with the values from corresponding default keyword
  coerceTypes: true, // change data type of data to match type keyword
  nullable: true // support keyword "nullable" from Open API 3 specification.
})
ajv.addFormat('binary', { type: 'string', validate: () => true })
ajv.addFormat('byte', { type: 'string', validate: () => true })
ajv.addFormat('int32', { type: 'number', validate: () => true })
ajv.addFormat('int64', { type: 'number', validate: () => true })
export function validatorCompiler ({ schema }: any): Ajv.ValidateFunction {
  return ajv.compile(schema)
}
```

The user can use the `validatorCompiler` in `route-level` or `instance-level` and they can even extend the format support.

### Instance Level Usage
```TS
import { validatorCompiler } from 'fastify-swagger`
fastify.setValidatorCompiler(validatorCompiler)
```

### Route Level Usage
```TS
import { validatorCompiler } from 'fastify-swagger'
fastify.get(
  '/',
  {
    schema: {
      body: {
        type: 'object',
        properties: { file: { type: 'string', format: 'binary' } }
      }
    },
    validatorCompiler
  },
  ( ) => {}
)
```

### Why I do not just add the custom ajv to fastify?
The main reason I do not add the custom `ajv` to `fastify` by option is that it will cause conflict to the user custom one. And it may break all the user when upgrading this package.

### PR DESCRIPTION
**Pull request by Adityapanther**
## issue fixed
#343
#47 

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: lib/validatorCompiler.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+const validatorCompiler = () => {
+  const ajv = new Ajv({
+    removeAdditional: true, // remove additional properties
+    useDefaults: true, // replace missing properties and items with the values from corresponding default keyword
+    coerceTypes: true, // change data type of data to match type keyword
+    nullable: true // support keyword "nullable" from Open API 3 specification.
+  })
+
+  ajv.addFormat('binary', {
+    type: 'string',
```
Comment (mcollina) [inline lib/validatorCompiler.js:12]: Is string representation correct?

File: lib/validatorCompiler.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    validate: () => true
+  })
+  ajv.addFormat('int32', {
+    type: 'number',
+    validate: () => true
+  })
+  ajv.addFormat('int64', {
+    type: 'number',
+    validate: () => true
+  })
```
Comment (mcollina) [inline lib/validatorCompiler.js:26]: Can you please implement those validation rules?

File: package.json [hunk 3]
```diff
@@ -27,6 +27,10 @@
     {
       "name": "Matteo Collina",
       "email": "hello@matteocollina.com"
+    },
+    {
+      "name": "Aditya panther",
+      "email": "raj68518@gmail.com"
```
Comment (mcollina) [inline package.json:33]: Please remove this, it's reserved for maintainers.

File: index.js [hunk 4]
```diff
@@ -25,7 +36,14 @@ function fastifySwagger (fastify, opts, next) {
...
   fastify: '>=3.x',
   name: 'fastify-swagger'
 })
+
+module.exports.default = plugin
+
+module.exports = {
+  fastifySwagger: plugin,
+  validatorCompiler
+}
```
Comment (mcollina) [inline index.js:49]: please restore the module.export to be the plugin. You fan attach the other properties to
module.exports.

Could you also restore the require of all the tests?

**Review by mcollina — CHANGES_REQUESTED** [review]
Good work!

File: lib/validatorCompiler.js (hunk 1 above)
Comment (Adityapanther) [reply lib/validatorCompiler.js:12]: can you pleaser explain more

File: index.js (hunk 4 above)
Comment (Adityapanther) [reply index.js:49]: we have a default export for plugins if user don't want to have object

File: package.json (hunk 3 above)
Comment (Adityapanther) [reply package.json:33]: ok, I will

File: lib/validatorCompiler.js (hunk 2 above)
Comment (Adityapanther) [reply lib/validatorCompiler.js:26]: which one

File: index.js (hunk 4 above)
Comment (Adityapanther) [reply index.js:49]: have a look on line no 44 (index.js)

File: package.json (hunk 3 above)
Comment (Adityapanther) [reply package.json:33]: ok done !

File: lib/validatorCompiler.js (hunk 1 above)
Comment (Adityapanther) [reply lib/validatorCompiler.js:12]: are you talking about comments

File: lib/validatorCompiler.js (hunk 2 above)
Comment (kibertoad) [reply lib/validatorCompiler.js:26]: You are simply returning `true` in all cases. Could you please implement rules checking if value is actually what is expected by schema?

Comment (Adityapanther) [reply lib/validatorCompiler.js:26]: ok

File: index.js (hunk 4 above)
Comment (Adityapanther) [reply index.js:49]: ok, fixed

File: lib/validatorCompiler.js (hunk 2 above)
Comment (Adityapanther) [reply lib/validatorCompiler.js:26]: done

**Comment by Adityapanther** [comment]
No, I removed

**Comment by kibertoad** [comment]
Thank you! Can you add tests for new formats? Also CI is broken.

**Comment by Adityapanther** [comment]
@kibertoad everything now good except CI, can you check CI because I have no experience with that or you can guide me for that

**Comment by kibertoad** [comment]
@Adityapanther See https://github.com/fastify/fastify-swagger/pull/458/checks?check_run_id=3455446735
tests are failing.

**Comment by Adityapanther** [comment]
ok, thanks

File: index.js [hunk 5]
```diff
@@ -25,7 +36,11 @@ function fastifySwagger (fastify, opts, next) {
   fastify.decorate('swaggerCSP', require('./static/csp.json'))
 }
 
-module.exports = fp(fastifySwagger, {
+const plugin = fp(fastifySwagger, {
   fastify: '>=3.x',
   name: 'fastify-swagger'
 })
+
+module.exports.default = plugin
```
Comment (climba03003) [inline index.js:44]: There is no need to specify `.default` or `.fastifySwagger`.
```suggestion
module.exports = plugin
module.exports.validatorCompiler = validatorCompiler
```

File: lib/validatorCompiler.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+  const ajv = ajvOpenApi(
+    new Ajv(ajvOptions),
+    openApiOptions
+  )
+  return ajv.compile(compileObj)(data)
+}
+
+const binaryValidation = (data) => {
+  return openAjv({ format: 'binary', type: 'string' }, data)
```
Comment (climba03003) [inline lib/validatorCompiler.js:26]: I means to take their validation as example and implement our own.

**Review by climba03003 — CHANGES_REQUESTED** [review]

File: index.js (hunk 5 above)
Comment (Adityapanther) [reply index.js:44]: I forgot to remove it, I will

File: lib/validatorCompiler.js (hunk 6 above)
Comment (Adityapanther) [reply lib/validatorCompiler.js:26]: ok, waitting for next weekend to work on that

File: index.js (hunk 5 above)
Comment (mcollina) [reply index.js:44]: default is needed: https://github.com/fastify/fastify/blob/a04f4c952731c57e79c6ca3794a62959a43731ed/fastify.js#L707-L717

File: lib/validatorCompiler.js (hunk 1 above)
Comment (mcollina) [reply lib/validatorCompiler.js:50]: It seems that a `binary` format would be represent by a binary type in JavaScript, but maybe I do not understand this correctly, could you please add a test?

File: test/validatorCompiler.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+            format: 'binary'
+          }
+        }
+      }
+    }
+
+  }, () => {})
+
+  fastify.ready(err => {
+    if (err) {
```
Comment (mcollina) [inline test/validatorCompiler.js:41]: no need for this `if (err)`, just `t.error(err)` does the trick.

File: index.js (hunk 5 above)
Comment (climba03003) [reply index.js:44]: @mcollina 
`.default` is not needed. It is added by `fastify-plugin` by default.
We do not need to re-export it.

// default export by fastify-plugin
https://github.com/fastify/fastify-plugin/blob/master/plugin.js#L10-L13
// named export by fastify-plugin
https://github.com/fastify/fastify-plugin/blob/master/plugin.js#L59-L62

File: test/validatorCompiler.js (hunk 7 above)
Comment (Adityapanther) [reply test/validatorCompiler.js:41]: ok

File: lib/validatorCompiler.js [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    schemaId: 'auto',
+    format: 'full',
+    coerceTypes: true,
+    unknownFormats: 'ignore',
+    useDefaults: true,
+    nullable: true
+  }
+
+  const openApiOptions = {
+    useDraft04: true
```
Comment (Eomm) [inline lib/validatorCompiler.js:15]: Is this needed?

File: test/validatorCompiler.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+            type: 'file',
+            format: 'binary'
+          }
+        }
+      }
+    }
+
+  }, () => {})
+
+  fastify.ready(err => {
```
Comment (Eomm) [inline test/validatorCompiler.js:40]: Could you call the endpoint?
We are not executing the validation function here so there could be hidden edge cases

File: index.js [hunk 10]
```diff
@@ -1,8 +1,19 @@
...
 
 const fp = require('fastify-plugin')
+const validatorCompiler = require('./lib/validatorCompiler')
 
 function fastifySwagger (fastify, opts, next) {
+  // enabling custom or validator complier form opts object
+  const customCompiler = opts.customCompiler || null
+  console.log(customCompiler)
+
+  if (customCompiler && customCompiler !== null && typeof customCompiler !== typeof undefined) {
```
Comment (Eomm) [inline index.js:11]: I would simply here

```suggestion
  if (customCompiler === true) {
```

**Review by Eomm — CHANGES_REQUESTED** [review]

File: lib/validatorCompiler.js [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+    nullable: true
+  })
+
+  ajv.addFormat('binary', {
+    type: 'string',
+    validate: binaryValidation
+  })
+  ajv.addFormat('byte', {
+    type: 'string',
+    validate: byteValidation
```
Comment (Eomm) [inline lib/validatorCompiler.js:55]: At every validation execution, we are creating a new ajv instance and compilation.
We must avoid it.

If you precompile the functions, it should work as well

File: lib/validatorCompiler.js (hunk 1 above)
Comment (Adityapanther) [reply lib/validatorCompiler.js:50]: @mcollina have a look on The formats defined by the OAS
https://swagger.io/specification/

File: index.js [hunk 12]
```diff
@@ -1,8 +1,19 @@
 'use strict'
 
 const fp = require('fastify-plugin')
+const validatorCompiler = require('./lib/validatorCompiler')
 
 function fastifySwagger (fastify, opts, next) {
+  // enabling custom or validator complier form opts object
+  const customCompiler = opts.customCompiler || null
+  console.log(customCompiler)
```
[the remark this answers is not in the document]
Comment (Adityapanther) [reply index.js:9]: done

File: test/validatorCompiler.js (hunk 9 above)
Comment (Adityapanther) [reply test/validatorCompiler.js:43]: can you please explain more ?

File: lib/validatorCompiler.js (hunk 11 above)
Comment (Adityapanther) [reply lib/validatorCompiler.js:72]: can you provide example

File: README.md [hunk 13]
```diff
@@ -122,6 +122,51 @@ fastify.ready(err => {
   fastify.swagger()
 })
 ```
+
+<a name="openapi.spec"></a>
+## OpenAPI spec (specification)
```
Comment (mcollina) [inline README.md:127]: I think this is not correct. It should be named "OpenAPI validation formats".

File: lib/validatorCompiler.js [hunk 14]
```diff
@@ -0,0 +1,92 @@
+'use strict'
+
+const Ajv = require('ajv')
+const Decimal = require('decimal.js')
```
Comment (mcollina) [inline lib/validatorCompiler.js:4]: Is this dependency actually needed?

File: README.md (hunk 13 above)
Comment (Adityapanther) [reply README.md:127]: okey

File: lib/validatorCompiler.js (hunk 14 above)
Comment (Adityapanther) [reply lib/validatorCompiler.js:4]: yes

File: test/validatorCompiler.js [hunk 15]
```diff
@@ -0,0 +1,104 @@
+const {
+  test
+} = require('tap')
+const Fastify = require('fastify')
+// const Swagger = require('openapi-schema-validator')
```
Comment (mcollina) [inline test/validatorCompiler.js:5]: why is this commented?

File: test/validatorCompiler.js (hunk 9 above)
Comment (mcollina) [reply test/validatorCompiler.js:43]: use fastify.inject to send an http request with the new formats.

File: test/validatorCompiler.js (hunk 15 above)
Comment (Adityapanther) [reply test/validatorCompiler.js:5]: it was not in used, so , removed it

**Comment by Adityapanther** [comment]
@mcollina  no need to use fastify.inject in test

**Comment by mcollina** [comment]
What this PR is missing to ship are a few tests, one for each of the new formats, that:

1. create an endpoint that uses the selected format, inside the route it validates what is the payload that is sent
2. performs a valid http request to the endpoint
3. performs a http request that would fail validation

Thanks

File: README.md [hunk 16]
```diff
@@ -207,6 +252,7 @@ An example of using `fastify-swagger` with `static` mode enabled can be found [h
  | Option             | Default          | Description                                                                                                               |
  | ------------------ | ---------------- | ------------------------------------------------------------------------------------------------------------------------- |
  | exposeRoute        | false            | Exposes documentation route.                                                                                              |
+ | customCompiler        | false  | if it's `true` means fastify will support swagger open api specs (binary, byte, int32, int64) as a schema                                                                                            |
```
Comment (Fdawgs) [inline README.md:255]:
```suggestion
 | customCompiler        | false  | If  `true` enable OpenAPI validation formats (binary, byte, int32, int64) as a schema                                                                                            |
```

File: README.md [hunk 17]
```diff
@@ -122,6 +122,51 @@ fastify.ready(err => {
   fastify.swagger()
 })
 ```
+
+<a name="openapi.validation"></a>
+## OpenAPI validation formats
+
+There are three way to enable open api specification
```
Comment (Fdawgs) [inline README.md:129]:
```suggestion
There are three way to enable the use of OpenAPI validation formats:
```

File: test/validatorCompiler.js [hunk 18]
```diff
@@ -0,0 +1,115 @@
+const {
+  test
+} = require('tap')
+const Fastify = require('fastify')
+// const Swagger = require('swagger-parser')
```
Comment (Fdawgs) [inline test/validatorCompiler.js:5]: Please remove commented out code if not needed

File: test/validatorCompiler.js [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+  // fastify.inject({
+  //   method: 'POST',
+  //   url: '/'
+  // }, (err, res) => {
+  //   t.error(err)
+  //   t.equal(res.statusCode, 200)
+  //   t.ok()
+  //   // const payload = JSON.parse(res.payload)
+  // })
+  // t.end()
```
Comment (Fdawgs) [inline test/validatorCompiler.js:114]: Please remove commented out code if not needed

File: README.md (hunk 16 above)
Comment (Fdawgs) [reply README.md:255]: This option also needs to be moved to be above "exposeRoute" as the table is sorted alphabetically ascending.

**Comment by Adityapanther** [comment]
I am currently busy with my company, I will update this sunday

**Comment by ec2ainun** [comment]
any update?

**Comment by CarterLi** [comment]
> any update?

He said he would update it this sunday. Lets be patient

**Comment by ec2ainun** [comment]
> > any update?
> 
> He said he would update it this sunday. Lets be patient

okay im sorry, i'm new here, but i think its already a month isn't it? @CarterLi

**Comment by CarterLi** [comment]
> > > any update?
> > 
> > 
> > He said he would update it this sunday. Lets be patient
> 
> okay im sorry, i'm new here, but i think its already a month isn't it? @CarterLi

I was joking...

**Comment by Adityapanther** [comment]
@ec2ainun I am quite busy in my life I am trying to manage my work and personal life, if anyone free here he can provide update to this pull otherwise I will provide update after my project finish

**Comment by ec2ainun** [comment]
@Adityapanther sorry, i have little knowledge about openapi spec implementation, i just look around how to create good documentation in order to upload file using fastify, and i found this pull request in discussion

**Comment by Adityapanther** [comment]
@mcollina  please review

File: test/validatorCompiler.js [hunk 20]
```diff
@@ -0,0 +1,126 @@
+const {
```
Comment (mcollina) [inline test/validatorCompiler.js:1]: add use strict at the top

**Review by mcollina — COMMENTED** [review]
It does not seem to include all the tests I asked before.

File: test/validatorCompiler.js (hunk 20 above)
Comment (Adityapanther) [reply test/validatorCompiler.js:1]: I added all as you asked

Comment (Adityapanther) [reply test/validatorCompiler.js:1]: please let me know If I left anything

Comment (climba03003) [reply test/validatorCompiler.js:1]: he means `"use strict";` for enable the strict mode.

Comment (Adityapanther) [reply test/validatorCompiler.js:1]: yes, but I think he is looking for anything else too

Comment (Adityapanther) [reply test/validatorCompiler.js:1]: I will add this tomorrow morning

**Comment by Adityapanther** [comment]
@mcollina added `use strict`

**Comment by mcollina** [comment]
> What this PR is missing to ship are a few tests, one for each of the new formats, that:
> 
> 1. create an endpoint that uses the selected format, inside the route it validates what is the payload that is sent
> 2. performs a valid http request to the endpoint
> 3. performs a http request that would fail validation
> 
> Thanks

I think these tests are still missing.
CI is also failing.
